src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java

```diff
@@ -0,0 +1,439 @@
+    private Map<BlockPos, WeakReference<IQuantumStorage<?>>> storageInstances = new HashMap<>();
+
+    /** The "definitive" set of positions of storage instances */
+    private Set<BlockPos> storagePositions = new HashSet<>();
+    private long energyConsumption = 0;
+    private final QuantumControllerHandler handler = new QuantumControllerHandler();
+
+    private boolean isDead = false;
+    private boolean isPowered = false;
+
+    public MetaTileEntityQuantumStorageController(ResourceLocation metaTileEntityId) {
+        super(metaTileEntityId);
+    }
+
+    @Override
+    public MetaTileEntity createMetaTileEntity(IGregTechTileEntity tileEntity) {
+        return new MetaTileEntityQuantumStorageController(metaTileEntityId);
+    }
+
+    @Override
+    public void update() {
+        super.update();
+        if (getWorld().isRemote) return;
+
+        if (getOffsetTimer() % 10 == 0) {
+            boolean isPowered = energyContainer.getEnergyStored() > energyConsumption && energyConsumption > 0;
+            if (isPowered) energyContainer.removeEnergy(energyConsumption);
+
+            if (isPowered != this.isPowered) {
+                this.isPowered = isPowered;
+                writeCustomData(GregtechDataCodes.UPDATE_ENERGY, buf -> buf.writeBoolean(this.isPowered));
+            }
+        }
+    }
+
+    public boolean isPowered() {
+        return isPowered;
+    }
+
+    @Override
+    public void receiveCustomData(int dataId, @NotNull PacketBuffer buf) {
+        super.receiveCustomData(dataId, buf);
+        if (dataId == GregtechDataCodes.UPDATE_ENERGY) {
+            this.isPowered = buf.readBoolean();
+            scheduleRenderUpdate();
+        } else if (dataId == GregtechDataCodes.UPDATE_ENERGY_PER) {
+            this.energyConsumption = buf.readLong();
+        }
+    }
+
+    @Override
+    public void renderMetaTileEntity(CCRenderState renderState, Matrix4 translation, IVertexOperation[] pipeline) {
+        var front = isPowered() ?
+                Textures.QUANTUM_CONTROLLER_FRONT_ACTIVE :
+                Textures.QUANTUM_CONTROLLER_FRONT_INACTIVE;
+        var sides = isPowered() ?
+                Textures.QUANTUM_CONTROLLER_ACTIVE :
+                Textures.QUANTUM_CONTROLLER_INACTIVE;
+
+        var newPipeline = ArrayUtils.add(pipeline,
+                new ColourMultiplier(GTUtility.convertRGBtoOpaqueRGBA_CL(getPaintingColorForRendering())));
+
+        front.renderSided(getFrontFacing(), renderState, translation, newPipeline);
+
+        for (EnumFacing facing : EnumFacing.VALUES) {
+            if (facing == getFrontFacing()) continue;
+            sides.renderSided(facing, renderState, translation, newPipeline);
+        }
+    }
+
+    @Override
+    public Pair<TextureAtlasSprite, Integer> getParticleTexture() {
+        return Pair.of(Textures.QUANTUM_CONTROLLER_ACTIVE.getParticleSprite(), getPaintingColorForRendering());
+    }
+
+    @Override
+    protected boolean openGUIOnRightClick() {
```

Comment:
UI could be cool, listing out all the connected inventories and their contents and positions. But it doesn't have to be implemented for this PR, since the network should work as is.

Reply:
this would be doable now that mui2 is merged into master (this pr was made before that)